### PR TITLE
Disallow referring to refs/remotes/origin/X by the "short name" of just X

### PR DESCRIPTION
**Pull request by PawelLipski**

### CODE REVIEW

File: backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepositoryFactory.java [hunk 1]
```diff
@@ -146,21 +146,33 @@ IGitMacheteRepository createGitMacheteRepository(IBranchLayout branchLayout) thr
 
       Map<String, List<IGitCoreReflogEntry>> filteredReflogByRemoteTrackingBranchName = remoteTrackingBranches
           .toMap(
-              /* keyMapper */ remoteBranch -> remoteBranch.getShortName(),
+              /* keyMapper */ remoteBranch -> remoteBranch.getName(),
```
Comment (PawelLipski) [inline backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepositoryFactory.java:149]: This line is the core of the fix

Comment (PawelLipski) [reply backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepositoryFactory.java:149]: "short name" used to be e.g. just `master` for `origin/master`, which clashed with `master` the local branch :frowning_face:

File: backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepositoryFactory.java [hunk 2]
```diff
@@ -146,21 +146,33 @@ IGitMacheteRepository createGitMacheteRepository(IBranchLayout branchLayout) thr
...
+      Seq<Tuple2<IGitCoreCommitHash, String>> commitHashAndBranchNamePairs = filteredReflogsByBranchName
+          .flatMap(branchNameAndReflog -> branchNameAndReflog._2
+              .map(re -> Tuple.of(re.getNewCommitHash(), branchNameAndReflog._1)));
+
+      var result = commitHashAndBranchNamePairs
+          .groupBy(commitHashAndBranchName -> commitHashAndBranchName._1)
+          .mapValues(pairsOfCommitHashAndBranchName -> pairsOfCommitHashAndBranchName
+              .map(commitHashAndBranchName -> commitHashAndBranchName._2));
+
+      LOG.debug(() -> "Derived the map of branches containing given commit in reflog:");
```
Comment (mkondratek) [inline backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepositoryFactory.java:171]: :+1:

File: backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepositoryFactory.java [hunk 3]
```diff
@@ -488,19 +500,19 @@ private SyncToRemoteStatus deriveSyncToRemoteStatus(IGitCoreLocalBranch coreLoca
         String newIdHash = e.getNewCommitHash().getHashString();
 
         if (e.getNewCommitHash().equals(entryToExcludeNewId)) {
-          LOG.trace(() -> "Exclude ${newIdHash} because it has the same hash as first entry");
+          LOG.trace(() -> "Exclude ${e} because it has the same hash as first entry");
```
Comment (mkondratek) [inline backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepositoryFactory.java:503]: What is the difference?

File: gitCoreApi/src/main/java/com/virtuslab/gitcore/api/IGitCoreBranch.java [hunk 4]
```diff
@@ -7,13 +7,19 @@
...
  * The only criterion for equality of any instances of any class implementing this interface is equality of
- * {@code derivePointedCommit} and {@code getFullName}
+ * {@code getFullName} and {@code derivePointedCommit}
  */
 public interface IGitCoreBranch {
   boolean isLocal();
 
-  String getShortName();
+  /**
+   * @return {@code X} part of {@code refs/heads/X} or {@code [remote-name]/X} part of {@code refs/heads/[remote-name]/X}
```
Comment (mkondratek) [inline gitCoreApi/src/main/java/com/virtuslab/gitcore/api/IGitCoreBranch.java:16]: :+1:

File: backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepositoryFactory.java (hunk 3 above)
Comment (PawelLipski) [reply backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepositoryFactory.java:503]: `e` is the entire entry, so not just new id but also old id and subject

**Pull request merged**